[SPARK-12293][SQL] Support UnsafeRow in LocalTableScan

viirya opened the pull request:
JIRA: https://issues.apache.org/jira/browse/SPARK-12293

Make LocalTableScan support UnsafeRow.

viirya committed:
Support UnsafeRow in LocalTableScan.

SparkQA commented:
**[Test build #47628 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47628/consoleFull)** for PR 10283 at commit [`ca4326a`](https://github.com/apache/spark/commit/ca4326ab0c4d26af3eddcc9550624db91d195e8a).

SparkQA commented:
**[Test build #47628 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47628/consoleFull)** for PR 10283 at commit [`ca4326a`](https://github.com/apache/spark/commit/ca4326ab0c4d26af3eddcc9550624db91d195e8a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class LocalRelation(output: Seq[Attribute], data: Seq[UnsafeRow] = Nil)`\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47628/
Test FAILed.

viirya committed:
Add copy().

SparkQA commented:
**[Test build #47634 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47634/consoleFull)** for PR 10283 at commit [`a0a991a`](https://github.com/apache/spark/commit/a0a991a7a0b0a8a17e3373ffc90edd74e64514e8).

SparkQA commented:
**[Test build #47634 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47634/consoleFull)** for PR 10283 at commit [`a0a991a`](https://github.com/apache/spark/commit/a0a991a7a0b0a8a17e3373ffc90edd74e64514e8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class LocalRelation(output: Seq[Attribute], data: Seq[UnsafeRow] = Nil)`\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47634/
Test FAILed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -937,8 +937,9 @@ object DecimalAggregates extends Rule[LogicalPlan] {
-      LocalRelation(projectList.map(_.toAttribute), data.map(projection))
+      val projection = UnsafeProjection.create(projectList, output)
+      LocalRelation(projectList.map(_.toAttribute),
+        data.map(projection(_).copy().asInstanceOf[UnsafeRow]))
```
The return type of `UnsafeProjection.apply` is `UnsafeRow` already, looks like we don't need the `asInstanceOf` here?

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala:
```diff
@@ -604,7 +606,9 @@ class SQLContext private[sql](
-    DataFrame(self, LocalRelation(attrSeq, rows.toSeq))
+    val projection = UnsafeProjection.create(attrSeq)
+    DataFrame(self,
+      LocalRelation(attrSeq, rows.toSeq.map(projection(_).copy().asInstanceOf[UnsafeRow])))
```
`LocalRelation.fromInternalRows`?

viirya committed:
Fix several bugs.

viirya commented:
I think the commits to fix these test errors are worth having their PR.

SparkQA commented:
**[Test build #47651 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47651/consoleFull)** for PR 10283 at commit [`f0e6ac0`](https://github.com/apache/spark/commit/f0e6ac0f412795f0d7ef5285eea0fc06a63dc270).

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:
```diff
@@ -61,6 +61,7 @@ object ScalaReflection extends ScalaReflection {
       case t if t <:< definitions.ByteTpe => ByteType
       case t if t <:< definitions.BooleanTpe => BooleanType
       case t if t <:< localTypeOf[Array[Byte]] => BinaryType
+      case t if t <:< localTypeOf[Decimal] => DecimalType.SYSTEM_DEFAULT
```
Normally `Decimal` should only be used inside spark SQL as the internal representation of decimal type, and we don't need to catch it here. Do we break it in tests?

viirya replied:
Yes. `constructorFor` will call `dataTypeFor` to determine if a type is `ObjectType` or not. If there is not case for `Decimal`, it will be recognized as `ObjectType` and causes bug.

SparkQA commented:
**[Test build #47651 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47651/consoleFull)** for PR 10283 at commit [`f0e6ac0`](https://github.com/apache/spark/commit/f0e6ac0f412795f0d7ef5285eea0fc06a63dc270).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class LocalRelation(output: Seq[Attribute], data: Seq[UnsafeRow] = Nil)`\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47651/
Test FAILed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:
```diff
@@ -421,46 +439,75 @@ object ScalaReflection extends ScalaReflection {
+                IsNull(unwrapped),
+                expressions.Literal.create(null, silentSchemaFor(optType).dataType),
+                Invoke(
+                  unwrapped,
```
If the `unwrapped` is null, the `Invoke` will return null, see https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects.scala#L172-L175

I think we don't need this extra `If`, did I miss something here?

viirya replied:
I think you are right. I add this to make sure it is not possibly creating wrong values when I debug. I will remove these `If`.

viirya committed:
Merge remote-tracking branch 'upstream/master' into unsaferow-localtablescan

viirya committed:
Fix remaining failed tests.

SparkQA commented:
**[Test build #47829 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47829/consoleFull)** for PR 10283 at commit [`97a81c3`](https://github.com/apache/spark/commit/97a81c3ab8a456968ed22e8a4af8e0ad8a505f26).

SparkQA commented:
**[Test build #47829 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47829/consoleFull)** for PR 10283 at commit [`97a81c3`](https://github.com/apache/spark/commit/97a81c3ab8a456968ed22e8a4af8e0ad8a505f26).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class LocalRelation(output: Seq[Attribute], data: Seq[UnsafeRow] = Nil)`\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47829/
Test FAILed.

viirya committed:
Merge remote-tracking branch 'upstream/master' into unsaferow-localtablescan

viirya committed:
Ignore the test for Limit node as it accepts UnsafeRow now.

SparkQA commented:
**[Test build #47888 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47888/consoleFull)** for PR 10283 at commit [`97d390f`](https://github.com/apache/spark/commit/97d390f9d4a929f6e115529d8bf8ec096074543c).

SparkQA commented:
**[Test build #47888 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47888/consoleFull)** for PR 10283 at commit [`97d390f`](https://github.com/apache/spark/commit/97d390f9d4a929f6e115529d8bf8ec096074543c).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47888/
Test FAILed.

viirya commented:
retest this please.

SparkQA commented:
**[Test build #47911 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47911/consoleFull)** for PR 10283 at commit [`97d390f`](https://github.com/apache/spark/commit/97d390f9d4a929f6e115529d8bf8ec096074543c).

SparkQA commented:
**[Test build #47911 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47911/consoleFull)** for PR 10283 at commit [`97d390f`](https://github.com/apache/spark/commit/97d390f9d4a929f6e115529d8bf8ec096074543c).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class LocalRelation(output: Seq[Attribute], data: Seq[UnsafeRow] = Nil)`\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47911/
Test FAILed.

viirya commented:
retest this please.

SparkQA commented:
**[Test build #47920 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47920/consoleFull)** for PR 10283 at commit [`97d390f`](https://github.com/apache/spark/commit/97d390f9d4a929f6e115529d8bf8ec096074543c).

SparkQA commented:
**[Test build #47920 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47920/consoleFull)** for PR 10283 at commit [`97d390f`](https://github.com/apache/spark/commit/97d390f9d4a929f6e115529d8bf8ec096074543c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class LocalRelation(output: Seq[Attribute], data: Seq[UnsafeRow] = Nil)`\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47920/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LocalRelation.scala:
```diff
@@ -29,20 +30,28 @@ object LocalRelation {
+  def fromProduct[T <: Product : ExpressionEncoder](
+      output: Seq[Attribute],
+      data: Seq[T]): LocalRelation = {
+    val encoder = implicitly[ExpressionEncoder[T]]
```
we can use `encoderFor[T]` here

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/GenericArrayData.scala:
```diff
@@ -18,14 +18,30 @@
 class GenericArrayData(val array: Array[Any]) extends ArrayData {
 
-  def this(seq: Seq[Any]) = this(seq.toArray)
+  def this(seq: Seq[Any]) = this(GenericArrayData.processSeq(seq))
```
instead of handling null here, I think a better way is not passing null to it.

viirya replied:
Thanks. I propose the fixing as pr #10401.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:
```diff
@@ -406,11 +419,16 @@ object ScalaReflection extends ScalaReflection {
-      if (isNativeType(catalystType)) {
-        NewInstance(
+
+      if (isNativeType(catalystType) && !(elementType <:< localTypeOf[Option[_]])) {
```
good catch!
The original one is wrong, how about `isNativeType(externalDataType)` which is simpler?

viirya replied:
Yes. Thanks. I propose this fixing as another pr #10391.

viirya commented:
SQLUserDefinedType support for encoder is proposed in another pr #10390.

viirya committed:
For comment.

viirya commented:
@cloud-fan I think I have addressed all your comments. These bugs found in implementing UnsafeRow support in LocalTableScan are submitted as other PRs with their tests, so you can review them better.

SparkQA commented:
**[Test build #48094 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48094/consoleFull)** for PR 10283 at commit [`2500de3`](https://github.com/apache/spark/commit/2500de3ba716ad93dca8001f5fde6c670c898416).

SparkQA commented:
**[Test build #48094 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48094/consoleFull)** for PR 10283 at commit [`2500de3`](https://github.com/apache/spark/commit/2500de3ba716ad93dca8001f5fde6c670c898416).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class LocalRelation(output: Seq[Attribute], data: Seq[UnsafeRow] = Nil)`\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48094/
Test FAILed.

viirya commented:
retest this please.

SparkQA commented:
**[Test build #48101 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48101/consoleFull)** for PR 10283 at commit [`2500de3`](https://github.com/apache/spark/commit/2500de3ba716ad93dca8001f5fde6c670c898416).

SparkQA commented:
**[Test build #48101 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48101/consoleFull)** for PR 10283 at commit [`2500de3`](https://github.com/apache/spark/commit/2500de3ba716ad93dca8001f5fde6c670c898416).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class LocalRelation(output: Seq[Attribute], data: Seq[UnsafeRow] = Nil)`\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48101/
Test PASSed.
